R/get_ref.R

```diff
@@ -285,22 +285,52 @@ get_ref_max <- function(remote_ref) {
 #' @importFrom pkgdepends parse_pkg_ref
 #' @export
 get_ref_release <- function(remote_ref) {
+  # create list of ref candidates to check
+  # return the one of the highest version
+  # this is a named list of character with version values and refs names
+  ref_candidates <- list()
   if (check_if_on_cran(remote_ref)) {
-    return(pkgdepends::parse_pkg_ref(remote_ref$package))
+    cran_ref <- remote_ref$package
+    cran_ver <- get_version(pkgdepends::parse_pkg_ref(cran_ref))
+    ref_candidates <- append(ref_candidates, setNames(list(cran_ver), cran_ref))
   }
   if (inherits(remote_ref, "remote_ref_github")) {
+    gh_release_ref <- cond_parse_pkg_ref_release(remote_ref)
+    if (!is.null(gh_release_ref)) {
+      gh_release_ver <- get_version(gh_release_ref)
+      ref_candidates <- c(ref_candidates, setNames(list(gh_release_ver), gh_release_ref$ref))
+    }
+
     if (!is.null(remote_ref$commitish) && remote_ref$commitish != "") {
-      return(remote_ref)
+      gh_commitish_ref <- remote_ref
+      gh_commitish_ver <- get_version(gh_commitish_ref)
+      ref_candidates <- c(ref_candidates, setNames(list(gh_commitish_ver), gh_commitish_ref$ref))
     }
     if (!is.null(remote_ref$pull) && remote_ref$pull != "") {
-      return(remote_ref)
+      gh_pull_ref <- remote_ref
+      gh_pull_ver <- get_version(gh_pull_ref)
+      ref_candidates <- c(ref_candidates, setNames(list(gh_pull_ver), gh_pull_ref$ref))
     }
-    if (!is.null(remote_ref$release) && remote_ref$release != "") {
-      return(cond_parse_pkg_ref_release(remote_ref))
+  } else {
+    input_ref <- remote_ref$ref
+    input_ver <- get_version(remote_ref)
+    ref_candidates <- c(ref_candidates, setNames(list(input_ver), input_ref))
+  }
+
+  if (length(ref_candidates) == 0 || all(is.na(ref_candidates))) {
+    return(remote_ref)
+  } else {
+    max_ver <- ref_candidates[[1]]
+    max_ref <- names(ref_candidates[1])
+    for (i in seq_along(ref_candidates)) {
+      i_ver <- ref_candidates[[i]]
+      i_ref <- names(ref_candidates[i])
+      if (!is.na(i_ver) && i_ver > max_ver) {
+        max_ref <- i_ref
+      }
     }
-    return(cond_parse_pkg_ref_release(remote_ref))
+    return(pkgdepends::parse_pkg_ref(max_ref))
   }
-  return(remote_ref)
 }
 
 #' @importFrom pkgdepends parse_pkg_ref
```

Comment:
WDYT about cleaning all these `@importFrom tags`? they seem unnecessary

I can do this on this PR if you agree

Reply:
yes - let's do that separately